compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12647,6 +12659,195 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+ *     ==>iRegLoad (cannotOverflow...)
+ *     iconst -1
+ *   ==>iconst 0
+ *
+ *
+ * After:
+ * ificmple --> block_17 BBStart at n13n ()
+ *   ==>iRegLoad (cannotOverflow...)
+ *   iconst 1
+ *
+ * \endverbatim
+ *
+ *
+ * The iadd node should be marked as 'cannotOverflow' to ensure functional correctness.
+ *
+ * \param node          the comparison node being simpilfied
+ * \param s             the simplifier object
+ * \param isInt         true if the integral types is less than or euqal to 32-bit wide;
+ *                      false if it's 64-bit.
+*/
+TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
+                                                TR::Simplifier * s,
+                                                bool isInt)
+   {
+   TR::Node* firstChild     = node->getFirstChild();
+   TR::Node* secondChild    = node->getSecondChild();
+   TR::ILOpCodes op         = firstChild->getOpCodeValue();
+
+   bool isAdd = (op == TR::iadd || op == TR::ladd || op == TR::iuadd || op == TR::luadd ||
+                 op == TR::sadd || op == TR::badd || op == TR::buadd);
+
+   bool isSub = (op == TR::isub || op == TR::lsub || op == TR::iusub || op == TR::lusub ||
+                 op == TR::ssub || op == TR::bsub || op == TR::busub);
+
+   if ((isAdd || isSub) &&
+           firstChild->chkCannotOverflow()  &&
+           firstChild->getSecondChild()->getOpCode().isLoadConst() &&
+           secondChild->getOpCode().isLoadConst() &&
+           firstChild->isSingleRef() &&
+           !(!firstChild->isSingleRef() &&
+             firstChild->getFutureUseCount() != firstChild->getReferenceCount() - 1))
+      {
+      bool isUnsigned = firstChild->getOpCode().isUnsigned();
+
+      int64_t const1 = 0, const2 = 0, const3 = 0;
+
+      // Get and check the integer/long constant values
+      if (isInt)
+         {
+         const1 = firstChild->getSecondChild()->getConstValue();
+         const2 = secondChild->getConstValue();
+         const3 = isAdd ? (const2 - const1) : (const2 + const1);
+
+         int64_t signedMax = 0, signedMin = 0;
+         uint64_t unsignedMax = 0, unsignedMin = 0;
+
+         if (firstChild->getOpCode().is1Byte())
+            {
+            signedMax   = !isUnsigned ? TR::getMaxSigned<TR::Int8>()   : 0;
+            signedMin   = !isUnsigned ? TR::getMinSigned<TR::Int8>()   : 0;
+            unsignedMax = isUnsigned  ? TR::getMaxUnsigned<TR::Int8>() : 0;
+            unsignedMin = isUnsigned  ? TR::getMinUnsigned<TR::Int8>() : 0;
+            }
+         else if (firstChild->getOpCode().is2Byte())
+            {
+            signedMax   = !isUnsigned ? TR::getMaxSigned<TR::Int16>()   : 0;
+            signedMin   = !isUnsigned ? TR::getMinSigned<TR::Int16>()   : 0;
+            unsignedMax = isUnsigned  ? TR::getMaxUnsigned<TR::Int16>() : 0;
+            unsignedMin = isUnsigned  ? TR::getMinUnsigned<TR::Int16>() : 0;
+            }
+         else if (firstChild->getOpCode().is4Byte())
```

Comment:
Smaller integers are extended to 64 bits for extra range so that we can check the true result. For integers that are already 64 bits wide, it would take an even wider integer type to do that